pkg/util/limiter/utilization.go

```diff
@@ -0,0 +1,330 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package limiter
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/dskit/services"
+	"github.com/pkg/errors"
+	"github.com/prometheus/procfs"
+	"go.uber.org/atomic"
+
+	"github.com/grafana/mimir/pkg/util/math"
+)
+
+const (
+	// Interval for updating resource (CPU/memory) utilization
+	resourceUtilizationUpdateInterval = time.Second
+
+	memPathV2 = "/sys/fs/cgroup/memory.current"
+	cpuPathV2 = "/sys/fs/cgroup/cpu.stat"
+	memPathV1 = "/sys/fs/cgroup/memory/memory.stat"
+	cpuPathV1 = "/sys/fs/cgroup/cpu/cpuacct.usage"
+)
+
+type utilizationScanner interface {
+	// Method returns the method for getting resource utilization.
+	Method() string
```

Comment:
Since this is just used for logging, maybe we could have implementations implement the Stringer interface instead?

Reply:
Done.